torch/lib/c10d/TCPStore.hpp

```diff
@@ -36,29 +70,55 @@ class TCPStoreDaemon {
   void getNumKeysHandler(int socket) const;
   void deleteHandler(int socket);
   void waitHandler(int socket);
+  void watchHandler(int socket);
 
   bool checkKeys(const std::vector<std::string>& keys) const;
+  // Helper function to alerts waiting workers, used in setHandler, getHandler
   void wakeupWaitingClients(const std::string& key);
-
-  void initStopSignal();
-  void closeStopSignal();
-
-  std::thread daemonThread_;
+  // Helper function used when the key is changed
+  // used in setHandler, addHandler, getHandler, deleteHandler
+  void sendKeyUpdatesToClients(
+      const std::string& key,
+      const enum WatchResponseType& type,
+      std::vector<uint8_t>& oldData,
+      std::vector<uint8_t>& newData);
   std::unordered_map<std::string, std::vector<uint8_t>> tcpStore_;
-  // From key -> the list of sockets waiting on it
+  // From key -> the list of sockets waiting on the key
   std::unordered_map<std::string, std::vector<int>> waitingSockets_;
   // From socket -> number of keys awaited
   std::unordered_map<int, size_t> keysAwaited_;
+  // From key -> the list of sockets watching the key
+  std::unordered_map<std::string, std::vector<int>> watchedSockets_;
+};
 
-  std::vector<int> sockets_;
-  int storeListenSocket_;
-#ifdef _WIN32
-  const std::chrono::milliseconds checkTimeout_
-      = std::chrono::milliseconds(10);
-  HANDLE ghStopEvent_;
-#else
-  std::vector<int> controlPipeFd_{-1, -1};
-#endif
+// Separate thread that is launched on all instances (including master)
+// Right now only handles callbacks registered from watchKey()
+class TCPStoreWorkerDaemon : public BackgroundThread {
+ public:
+  explicit TCPStoreWorkerDaemon(int listenSocket);
+  // Adds a callback to run key change
+  void addCallback(
```

Comment:
curious, why do we need `addCallback` and `watchKey` as two separate APIs? Do we expect users directly call `addCallback`? Or is this mainly for testing purpose?

Reply:
the map from key -> callback is stored on TCPStoreWorkerDaemon and `addCallback` is a helper method to add to it. `watchKey` is what registers the callback (add socket to TCPStoreMasterDaemon and callback to TCPStoreWorkerDaemon) so `watchKey` is the only API that should be called

Reply:
I see, does this mean `addCallback` should be private?

Reply:
I wanted to keep the keyToCallback_ map private in TCPWorkerDaemon but still need to update it from the user process, so I need the method to be public. I will rename it to `setCallback` to be clearer